master/internal/token/postgres_token.go

```diff
@@ -0,0 +1,160 @@
+package token
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/o1egl/paseto"
+	"github.com/uptrace/bun"
+	"gopkg.in/guregu/null.v3"
+
+	"github.com/determined-ai/determined/master/internal/db"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+const (
+	// DefaultTokenLifespan is how long a newly created access token is valid.
+	DefaultTokenLifespan = 30 * 24 * time.Hour
+)
+
+// CurrentTimeNowInUTC stores the current time in UTC for time insertions.
+var CurrentTimeNowInUTC time.Time
```

Comment:
I think we should do away with this variable because it's kind of a footgun that might lead to race conditions—in general, shared state between threads or goroutines that's not behind some sort of thread safety mechanism like a mutex is a race condition danger. Since different goroutines can call `CreateAccessToken`, this variable isn't safe.

I think all references to this can be replaced with a simple `time.Now().UTC()`. `WithTokenExpiry` probably needs to be a tad more complex and call `time.Now().UTC()` and then return a closure that closes over that value or something like this:

```go
// WithTokenExpiry function will add specified expiresAt (if any) to the access token table.
func WithTokenExpiry(expiry *time.Duration) AccessTokenOption {
	now := time.Now().UTC()
	return func(s *model.UserSession) {
		s.Expiry = now.Add(*expiry)
	}
}
```

Or possibly that `time.Now().UTC()` can be called within the `AccessTokenOption` function returned:
```go
// WithTokenExpiry function will add specified expiresAt (if any) to the access token table.
func WithTokenExpiry(expiry *time.Duration) AccessTokenOption {
	return func(s *model.UserSession) {
		now := time.Now().UTC()
		s.Expiry = now.Add(*expiry)
	}
}
```
Sorry, I'm in a rush and am trying to get this feedback to you asap, so I haven't thought it fully through yet to know which of these is more correct. You may know quickly, though. Feel free to reach out on Slack if you want to talk more about it quickly.

Reply:
So you mean - This `CurrentTimeNowInUTC`  variable is mutable and shared across all goroutines calling `CreateAccessToken`. If two goroutines call this function concurrently, one goroutine might overwrite `CurrentTimeNowInUTC` while the other is still using it. If one goroutine modifies `CurrentTimeNowInUTC`, it can lead to incorrect behavior for another goroutine still working with that shared value, leading to unpredictable outcomes (like incorrect token expiration times).

Only reason for me to do this was - By using `CurrentTimeNowInUTC` for all time-related operations within `CreateAccessToken`, the code ensured that all timestamps (e.g., `CreatedAt`, `Expiry`, etc.) were based on the same moment in time. Otherwise I noted that there was a nanoseconds difference between them.

Let me know if the below refactored code makes sense - 

```
// WithTokenExpiry adds the specified expiresAt (if any) to the access token table.
func WithTokenExpiry(expiry *time.Duration) AccessTokenOption {
    // Calculate the current time within the closure, avoiding shared state.
    return func(s *model.UserSession) {
        now := time.Now().UTC()  // Call time.Now().UTC() directly here.
        s.Expiry = now.Add(*expiry)
    }
}

....

// CreateAccessToken creates a new access token and stores it in the user_sessions db.
func CreateAccessToken(
    ctx context.Context, userID model.UserID, opts ...AccessTokenOption,
) (string, model.TokenID, error) {
    now := time.Now().UTC()  // Call time.Now().UTC() at the point of token creation.

    // Populate the default values in the model.
    accessToken := &model.UserSession{
        UserID:      userID,
        CreatedAt:   now,
        Expiry:      now.Add(DefaultTokenLifespan),
        TokenType:   model.TokenTypeAccessToken,
        Description: null.StringFromPtr(nil),
        RevokedAt:   null.Time{},
    }

    // Apply any provided options (e.g., WithTokenExpiry, WithTokenDescription)
    for _, opt := range opts {
        opt(accessToken)
    }

```

TLDR, Remove the global `CurrentTimeNowInUTC` and replace all its references with `time.Now().UTC()`.
Update the `WithTokenExpiry` function so that it calculates the current time locally, and uses that for setting the token's expiry.

Reply:
Yes, that's exactly what I mean—you nailed it :100:. And yes, I see what you mean where if you used `time.Now().UTC()` directly everywhere rather than saving a local variable, they'd be slightly off from each other.

And yeah, your updated code looks great!